extension/extensionlimiter/extensionlimiter.go

```diff
@@ -0,0 +1,116 @@
+
+	// WeightKeyMemorySize is typically used with ResourceLimiters
+	// for limiting active memory usage.
+	WeightKeyMemorySize WeightKey = "memory_size"
+)
+
+// Provider is an interface that provides access to different limiter types
+// for specific weight keys.
+//
+// Extensions implementing this interface can be referenced by their
+// names from component rate limiting configurations (e.g., limitermiddleware).
+type Provider interface {
+	// RateLimiter returns a RateLimiter for the specified weight key.
+	//
+	// Rate limiters are useful for limiting an amount of
+	// something per unit of time.  In the OpenTelemetry metrics
+	// data model, resource limiters would apply to value of an
+	// Counter, where the value is incremented on Limit().
+	RateLimiter(key WeightKey) RateLimiter
+
+	// ResourceLimiter returns a ResourceLimiter for the specified
+	// weight key.
+	//
+	// Resource limiters are useful for limiting a total amount of
+	// something.  In the OpenTelemetry metrics data model,
+	// resource limiters would apply to value of an UpDownCounter,
+	// where the value is incremented on Acquire() and decremented
+	// in Release().
+	//
+	// In cases where a component supports a rate limiter and does not use
+	// a release function, the component may return a ResourceLimiterFunc
+	// which calls the underlying rate limiter and returns a no-op ReleaseFunc.
+	ResourceLimiter(key WeightKey) ResourceLimiter
+}
+
+// ResourceLimiter is an interface that components can use to apply
+// resource limiting (e.g., concurrent requests, memory in use).
+type ResourceLimiter interface {
+	// Acquire attempts to acquire resources based on the provided weight value.
+	//
+	// It may block until resources are available or return an error if the limit
+	// cannot be satisfied.
+	//
+	// On success, it returns a ReleaseFunc that should be called
+	// when the resources are no longer needed.
+	Acquire(ctx context.Context, value uint64) (ReleaseFunc, error)
+}
+
+var _ ResourceLimiter = ResourceLimiterFunc(nil)
+
+// ReleaseFunc is called when resources should be released after limiting.
+//
+// RelaseFunc values are never nil values, even in the error case, for
+// safety. Users should unconditionally defer these.
+type ReleaseFunc func()
+
+// ResourceLimiterFunc is an easy way to construct ResourceLimiters.
+type ResourceLimiterFunc func(ctx context.Context, value uint64) (ReleaseFunc, error)
+
+// Acquire implements ResourceLimiter.
+func (f ResourceLimiterFunc) Acquire(ctx context.Context, value uint64) (ReleaseFunc, error) {
+	if f == nil {
+		return func() {}, nil
+	}
+	return f(ctx, value)
+}
+
+// RateLimiter is an interface that components can use to apply
+// rate limiting (e.g., network-bytes-per-second, requests-per-second).
+type RateLimiter interface {
+	// Limit attempts to apply rate limiting based on the provided weight value.
+	// Limit is expected to block the caller until the weight can be admitted.
```

Comment:
This is very confusing:
I see that https://github.com/open-telemetry/opentelemetry-collector/pull/12700/files#diff-50228098c7b1cb3e86d349cf309bfc70dae29f56dc0374f9a5e432f9659cf926R38 returns if memory is over limit immediately which in my opinion is the right thing.

Here `Limit is expected to block the caller until the weight can be admitted.`, I don't agree with this.